Panic in HTTP push handler

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0xf4af36]

goroutine 24705 [running]:
github.com/livepeer/go-livepeer/server.(*LivepeerServer).HandlePush.func1(0xc00016de00, 0xc000231b20, 0xc0001f45a0, 0x24)
        /build/server/mediaserver.go:624 +0xd6
created by github.com/livepeer/go-livepeer/server.(*LivepeerServer).HandlePush
        /build/server/mediaserver.go:621 +0x8a9

--- a/server/mediaserver.go
+++ b/server/mediaserver.go
@@ -54,11 +54,12 @@ const StreamKeyBytes = 6
 
 const SegLen = 2 * time.Second
 const BroadcastRetry = 15 * time.Second
-const RefreshIntervalHttpPush = 1 * time.Minute
 
 var BroadcastJobVideoProfiles = []ffmpeg.VideoProfile{ffmpeg.P240p30fps4x3, ffmpeg.P360p30fps16x9}
 
 var AuthWebhookURL string
+
+var refreshIntervalHttpPush = 1 * time.Minute
 
 type streamParameters struct {
 	mid        core.ManifestID
@@ -410,9 +411,10 @@ func (s *LivepeerServer) registerConnection(rtmpStrm stream.RTMPVideoStream) (*r
 		Bitrate:    "4000k", // Fix this
 	}
 	hlsStrmID := core.MakeStreamID(mid, &vProfile)
-	s.connectionLock.Lock()
+	s.connectionLock.RLock()
+	// Fast path - check early if session exists - creating new session can take time
 	_, exists := s.rtmpConnections[mid]
-	s.connectionLock.Unlock()
+	s.connectionLock.RUnlock()
 	if exists {
 		// We can only have one concurrent stream per ManifestID
 		return nil, errAlreadyExists
@@ -431,6 +433,14 @@ func (s *LivepeerServer) registerConnection(rtmpStrm stream.RTMPVideoStream) (*r
 	}
 
 	s.connectionLock.Lock()
+	_, exists = s.rtmpConnections[mid]
+	// Check if session exist again - potentially two sessions can be created simultaneously,
+	// so we don't want to overwrite one that was already created
+	if exists {
+		// We can only have one concurrent stream per ManifestID
+		s.connectionLock.Unlock()
+		return nil, errAlreadyExists
+	}
 	s.rtmpConnections[mid] = cxn
 	s.lastManifestID = mid
 	s.lastHLSStreamID = hlsStrmID
@@ -592,6 +602,10 @@ func (s *LivepeerServer) HandlePush(w http.ResponseWriter, r *http.Request) {
 
 	now := time.Now()
 	mid := parseManifestID(r.URL.Path)
+	if mid == "" {
+		http.Error(w, `Bad URL`, http.StatusBadRequest)
+		return
+	}
 	s.connectionLock.Lock()
 	cxn, exists := s.rtmpConnections[mid]
 	if exists && cxn != nil {
@@ -616,17 +630,19 @@ func (s *LivepeerServer) HandlePush(w http.ResponseWriter, r *http.Request) {
 			return
 		}
 
-		ticker := time.NewTicker(RefreshIntervalHttpPush)
+		ticker := time.NewTicker(refreshIntervalHttpPush)
 
 		go func(s *LivepeerServer, mid core.ManifestID) {
+			defer ticker.Stop()
 			for range ticker.C {
+				var lastUsed time.Time
 				s.connectionLock.RLock()
-				lastUsed := s.rtmpConnections[mid].lastUsed
+				if cxn, exists := s.rtmpConnections[mid]; exists {
+					lastUsed = cxn.lastUsed
+				}
 				s.connectionLock.RUnlock()
-
-				if time.Since(lastUsed) > RefreshIntervalHttpPush {
+				if time.Since(lastUsed) > refreshIntervalHttpPush {
 					_ = removeRTMPStream(s, mid)
-					ticker.Stop()
 					return
 				}
 			}

--- a/server/push_test.go
+++ b/server/push_test.go
@@ -8,6 +8,7 @@ import (
 	"os"
 	"strings"
 	"testing"
+	"time"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
@@ -56,6 +57,86 @@ func TestMemoryRequestError(t *testing.T) {
 	require.Nil(t, err)
 	assert.Equal(http.StatusInternalServerError, resp.StatusCode)
 	assert.Contains(strings.TrimSpace(string(body)), "Error reading http request body")
+}
+
+func TestEmptyURLError(t *testing.T) {
+	// assert http request body error returned
+	assert := assert.New(t)
+	s := setupServer()
+	w := httptest.NewRecorder()
+	req := httptest.NewRequest("POST", "/live/.ts", nil)
+	s.HandlePush(w, req)
+	resp := w.Result()
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	require.Nil(t, err)
+	assert.Equal(http.StatusBadRequest, resp.StatusCode)
+	assert.Equal("Bad URL\n", string(body))
+}
+
+func TestShouldUpdateLastUsed(t *testing.T) {
+	assert := assert.New(t)
+	s := setupServer()
+	w := httptest.NewRecorder()
+	req := httptest.NewRequest("POST", "/live/mani1/1.ts", nil)
+	s.HandlePush(w, req)
+	resp := w.Result()
+	resp.Body.Close()
+	lu := s.rtmpConnections["mani1"].lastUsed
+	w = httptest.NewRecorder()
+	req = httptest.NewRequest("POST", "/live/mani1/1.ts", nil)
+	s.HandlePush(w, req)
+	resp = w.Result()
+	resp.Body.Close()
+	assert.True(lu.Before(s.rtmpConnections["mani1"].lastUsed))
+}
+
+func TestShouldRemoveSessionAfterTimeout(t *testing.T) {
+	oldRI := refreshIntervalHttpPush
+	refreshIntervalHttpPush = 2 * time.Millisecond
+	assert := assert.New(t)
+	s := setupServer()
+	w := httptest.NewRecorder()
+	req := httptest.NewRequest("POST", "/live/mani3/1.ts", nil)
+	s.HandlePush(w, req)
+	resp := w.Result()
+	resp.Body.Close()
+	s.connectionLock.Lock()
+	_, exists := s.rtmpConnections["mani3"]
+	s.connectionLock.Unlock()
+	assert.True(exists)
+	time.Sleep(50 * time.Millisecond)
+	s.connectionLock.Lock()
+	_, exists = s.rtmpConnections["mani3"]
+	s.connectionLock.Unlock()
+	assert.False(exists)
+	refreshIntervalHttpPush = oldRI
+}
+
+func TestShouldNotPanicIfSessionAlreadyRemoved(t *testing.T) {
+	oldRI := refreshIntervalHttpPush
+	refreshIntervalHttpPush = 5 * time.Millisecond
+	assert := assert.New(t)
+	s := setupServer()
+	w := httptest.NewRecorder()
+	req := httptest.NewRequest("POST", "/live/mani2/1.ts", nil)
+	s.HandlePush(w, req)
+	resp := w.Result()
+	resp.Body.Close()
+	s.connectionLock.Lock()
+	_, exists := s.rtmpConnections["mani2"]
+	s.connectionLock.Unlock()
+	assert.True(exists)
+	s.connectionLock.Lock()
+	delete(s.rtmpConnections, "mani2")
+	s.connectionLock.Unlock()
+	time.Sleep(10 * time.Millisecond)
+	s.connectionLock.Lock()
+	_, exists = s.rtmpConnections["mani2"]
+	s.connectionLock.Unlock()
+	assert.False(exists)
+	refreshIntervalHttpPush = oldRI
 }
 
 func TestFileExtensionError(t *testing.T) {

